common/quotas/global/collection/internal/atomicmap.go

```diff
@@ -0,0 +1,104 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package internal
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// AtomicMap adds type safety around a sync.Map (which has atomic-like behavior), and:
+//   - implicitly constructs values as needed, not relying on zero values
+//   - simplifies the API quite a bit because very few methods are in use.
+//     in particular there is no "Store" currently because it is not needed.
+//   - tracks length (atomically, so values are only an estimate)
```

Comment:
is the length only an estimate number?

Reply:
it's mostly a "by the time the func returns it may be incorrect, so you can never trust it except as an estimate".

the count should be accurate / it won't drift or be a `return 100` false number of some kind.  but it's inherently unreliable.